
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.3, mkdocs-material-7.3.6">
    
    
      
        <title>Genomic Range Indexing - Genomics Extension for SQLite</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.a57b2b03.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.3f5d1f46.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
    
      


    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    <script>function __prefix(e){return new URL("..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#programming-guide-genomic-range-indexing" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Genomics Extension for SQLite" class="md-header__button md-logo" aria-label="Genomics Extension for SQLite" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Genomics Extension for SQLite
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Genomic Range Indexing
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Genomics Extension for SQLite" class="md-nav__button md-logo" aria-label="Genomics Extension for SQLite" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    Genomics Extension for SQLite
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Intro & Install
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../guide_db/" class="md-nav__link">
        Opening Compressed Databases
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Genomic Range Indexing
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Genomic Range Indexing
      </a>
      
        


<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#conventions" class="md-nav__link">
    Conventions
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#create-gri" class="md-nav__link">
    Create GRI
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#query-gri" class="md-nav__link">
    Query GRI
  </a>
  
    <nav class="md-nav" aria-label="Query GRI">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#level-bounds-optimization" class="md-nav__link">
    Level bounds optimization
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#joining-tables-on-range-overlap" class="md-nav__link">
    Joining tables on range overlap
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reference-genome-metadata" class="md-nav__link">
    Reference genome metadata
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cookbook" class="md-nav__link">
    Cookbook
  </a>
  
    <nav class="md-nav" aria-label="Cookbook">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#rid-to-chromosome-name" class="md-nav__link">
    rid to chromosome name
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#query-rid-using-chromosome-name" class="md-nav__link">
    Query rid using chromosome name
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#circular-chromosome-query" class="md-nav__link">
    Circular chromosome query
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#advice-for-big-data" class="md-nav__link">
    Advice for big data
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../guide_helpers/" class="md-nav__link">
        Useful Routines
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../bindings/" class="md-nav__link">
        Writing Language Bindings
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../internals/" class="md-nav__link">
        Internals
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#conventions" class="md-nav__link">
    Conventions
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#create-gri" class="md-nav__link">
    Create GRI
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#query-gri" class="md-nav__link">
    Query GRI
  </a>
  
    <nav class="md-nav" aria-label="Query GRI">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#level-bounds-optimization" class="md-nav__link">
    Level bounds optimization
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#joining-tables-on-range-overlap" class="md-nav__link">
    Joining tables on range overlap
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reference-genome-metadata" class="md-nav__link">
    Reference genome metadata
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cookbook" class="md-nav__link">
    Cookbook
  </a>
  
    <nav class="md-nav" aria-label="Cookbook">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#rid-to-chromosome-name" class="md-nav__link">
    rid to chromosome name
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#query-rid-using-chromosome-name" class="md-nav__link">
    Query rid using chromosome name
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#circular-chromosome-query" class="md-nav__link">
    Circular chromosome query
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#advice-for-big-data" class="md-nav__link">
    Advice for big data
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                
                <h1 id="programming-guide-genomic-range-indexing">Programming Guide - Genomic Range Indexing</h1>
<p>GenomicSQLite enables creation of a <strong>Genomic Range Index (GRI)</strong> for any database table in which each row represents a genomic feature with (chromosome, beginPosition, endPosition) coordinates. The coordinates may be sourced from table columns or by computing arithmetic expressions thereof. The index tracks any updates to the underlying table as usual, with one caveat explained below.</p>
<p>Once indexed, the table can be queried for all features overlapping a query range. A GRI query yields a <a href="https://www.sqlite.org/rowidtable.html">rowid</a> set, which your SQL query can select from the indexed table for further filtering or analysis. Please review the brief SQLite documentation on <a href="https://www.sqlite.org/rowidtable.html">rowid</a> and <a href="https://www.sqlite.org/autoinc.html">Autoincrement</a>.</p>
<h3 id="conventions">Conventions</h3>
<p>Range positions are considered <a href="http://www.cs.utexas.edu/users/EWD/transcriptions/EWD08xx/EWD831.html"><strong>zero-based &amp; half-open</strong></a>, so the length of a feature is exactly endPosition-beginPosition nucleotides. The implementation doesn't strictly require this convention, but we strongly recommend observing it to minimize confusion. There is no practical limit on chromosome length, as position values may go up to 2<sup>60</sup>, but queries have a runtime factor logarithmic in the maximum feature length.</p>
<p>The extension provides routines to populate a small <code>_gri_refseq</code> table describing the genomic reference sequences, which other tables can reference by integer ID ("rid") instead of storing a column with textual sequence names like 'chr10'. This convention is not required, as the GRI can index either chromosome name or rid columns, but reasons to observe it include:</p>
<ul>
<li>Integers are more compact and faster to look up.</li>
<li>Results sort properly with <code>ORDER BY rid</code> instead of considering e.g. <code>'chr10'</code> &lt; <code>'chr2'</code> lexicographically. (See also the UINT collating sequence, below)</li>
<li>A table with chromosome names can be reconstructed easily by joining with <code>_gri_refseq</code>.</li>
</ul>
<h3 id="create-gri">Create GRI</h3>
<p><strong>↪ Create Genomic Range Index SQL:</strong> <em>Generate a string</em> containing a series of SQL statements which <em>when executed</em> create a GRI on an existing table. Executing them is left to the caller, perhaps after logging the contents. The statements should be executed within a transaction to succeed or fail atomically.</p>
<div class="tabbed-set" data-tabs="1:6"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><label for="__tabbed_1_1">Python</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="n">create_gri_sql</span> <span class="o">=</span> <span class="n">genomicsqlite</span><span class="o">.</span><span class="n">create_genomic_range_index_sql</span><span class="p">(</span>
  <span class="n">dbconn</span><span class="p">,</span>
  <span class="s1">&#39;tableName&#39;</span><span class="p">,</span>
  <span class="s1">&#39;chromosome&#39;</span><span class="p">,</span>
  <span class="s1">&#39;beginPosition&#39;</span><span class="p">,</span>
  <span class="s1">&#39;endPosition&#39;</span>
<span class="p">)</span>
<span class="n">dbconn</span><span class="o">.</span><span class="n">executescript</span><span class="p">(</span><span class="n">create_gri_sql</span><span class="p">)</span>
</code></pre></div>
</div>
<input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><label for="__tabbed_1_2">Java</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="n">String</span> <span class="n">griSql</span> <span class="o">=</span> <span class="n">GenomicSQLite</span><span class="p">.</span><span class="na">createGenomicRangeIndexSQL</span><span class="p">(</span>
  <span class="n">dbconn</span><span class="p">,</span>
  <span class="s">&quot;tableName&quot;</span><span class="p">,</span>
  <span class="s">&quot;chromosome&quot;</span><span class="p">,</span>
  <span class="s">&quot;beginPosition&quot;</span><span class="p">,</span>
  <span class="s">&quot;endPosition&quot;</span>
<span class="p">);</span>
<span class="n">dbconn</span><span class="p">.</span><span class="na">createStatement</span><span class="p">().</span><span class="na">executeUpdate</span><span class="p">(</span><span class="n">griSql</span><span class="p">);</span>
</code></pre></div>
</div>
<input id="__tabbed_1_3" name="__tabbed_1" type="radio" /><label for="__tabbed_1_3">Rust</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="kd">let</span><span class="w"> </span><span class="n">gri_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dbconn</span><span class="w"></span>
<span class="w">  </span><span class="p">.</span><span class="n">create_genomic_range_index_sql</span><span class="p">(</span><span class="s">&quot;tableName&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;chromosome&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">                                  </span><span class="s">&quot;beginPosition&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;endPosition&quot;</span><span class="p">)</span><span class="o">?</span><span class="p">;</span><span class="w"></span>
<span class="n">dbconn</span><span class="p">.</span><span class="n">execute_batch</span><span class="p">(</span><span class="o">&amp;</span><span class="n">gri_sql</span><span class="p">)</span><span class="o">?</span><span class="p">;</span><span class="w"></span>
</code></pre></div>
</div>
<input id="__tabbed_1_4" name="__tabbed_1" type="radio" /><label for="__tabbed_1_4">SQLiteCpp</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="nf">CreateGenomicRangeIndexSQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">table</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">rid</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">beg</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">end</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="kt">int</span><span class="w"> </span><span class="n">floor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">-1</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="n">create_gri_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">CreateGenomicRangeIndexSQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="s">&quot;tableName&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;chromosome&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;beginPosition&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;endPosition&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>
<span class="c1">// SQLite::Database* dbconn in a transaction</span>
<span class="n">dbconn</span><span class="o">-&gt;</span><span class="n">exec</span><span class="p">(</span><span class="n">create_gri_sql</span><span class="p">);</span><span class="w"></span>
</code></pre></div>
</div>
<input id="__tabbed_1_5" name="__tabbed_1" type="radio" /><label for="__tabbed_1_5">C++</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="nf">CreateGenomicRangeIndexSQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">table</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">rid</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">beg</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">end</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="kt">int</span><span class="w"> </span><span class="n">floor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">-1</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="n">create_gri_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">CreateGenomicRangeIndexSQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="s">&quot;tableName&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;chromosome&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;beginPosition&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;endPosition&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>
<span class="c1">// sqlite3* dbconn in a transaction</span>
<span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">errmsg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">nullptr</span><span class="p">;</span><span class="w"></span>
<span class="kt">int</span><span class="w"> </span><span class="n">rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sqlite3_exec</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span><span class="w"> </span><span class="n">create_gri_sql</span><span class="p">.</span><span class="n">c_str</span><span class="p">(),</span><span class="w"> </span><span class="k">nullptr</span><span class="p">,</span><span class="w"> </span><span class="k">nullptr</span><span class="p">,</span><span class="w"> </span><span class="o">&amp;</span><span class="n">errmsg</span><span class="p">);</span><span class="w"></span>
<span class="c1">// check rc, free errmsg</span>
</code></pre></div>
</div>
<input id="__tabbed_1_6" name="__tabbed_1" type="radio" /><label for="__tabbed_1_6">C</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="nf">create_genomic_range_index_sql</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">table</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">rid</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">beg</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">end</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="kt">int</span><span class="w"> </span><span class="n">floor</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>
<span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">create_gri_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">create_genomic_range_index_sql</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="s">&quot;tableName&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;chromosome&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;beginPosition&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;endPosition&quot;</span><span class="p">,</span><span class="w"> </span><span class="mi">-1</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>
<span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">*</span><span class="n">create_gri_sql</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">errmsg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">0</span><span class="p">;</span><span class="w"></span>
<span class="w">  </span><span class="cm">/* sqlite3* dbconn in a transaction */</span><span class="w"></span>
<span class="w">  </span><span class="kt">int</span><span class="w"> </span><span class="n">rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sqlite3_exec</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span><span class="w"> </span><span class="n">create_gri_sql</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="o">&amp;</span><span class="n">errmsg</span><span class="p">);</span><span class="w"></span>
<span class="w">  </span><span class="cm">/* check rc, free errmsg */</span><span class="w"></span>
<span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w"> </span><span class="cm">/* General note: all GenomicSQLite C routines returning a char* string use</span>
<span class="cm">  * the following convention:</span>
<span class="cm">  * If the operation suceeds then it&#39;s a nonempty, null-terminated string.</span>
<span class="cm">  * Otherwise it points to a null byte followed immediately by a nonempty,</span>
<span class="cm">  * null-terminated error message.</span>
<span class="cm">  * IN EITHER CASE, the caller should free the string with sqlite3_free().</span>
<span class="cm">  * Null is returned only if malloc failed.</span>
<span class="cm">  */</span><span class="w"></span>
<span class="p">}</span><span class="w"></span>
<span class="n">sqlite3_free</span><span class="p">(</span><span class="n">create_gri_sql</span><span class="p">);</span><span class="w"></span>
</code></pre></div>
</div>
</div>
<p>The three arguments following the table name tell the indexing procedure how to read the feature coordinates from each table row.</p>
<ol>
<li>The reference sequence may be sourced either by the name of a text column containing names like 'chr10', or of an integer reference ID (rid) column, as discussed above.</li>
<li>The begin and end positions are read from named integer columns, or by computing simple arithmetic expressions thereof.</li>
<li>For example, if the table happens to have <code>beginPosition</code> and <code>featureLength</code> columns, the end position may be formulated <code>'beginPosition+featureLength'</code>.</li>
</ol>
<p><strong>❗ The table name and expressions are textually pasted into a template SQL script. Take care to prevent SQL injection, if they're in any way determined by external input.</strong></p>
<p>A last optional integer argument <code>floor</code> can be omitted or left at -1. <small>GRI performance may be improved slightly by setting <code>floor</code> to a positive integer <em>F</em> if the following is true: the lengths of the indexed features are almost all &gt;16<sup><em>F</em>-1</sup>, with only very few outlier lengths &le;16<sup><em>F</em>-1</sup>. For example, human exons are almost all &gt;16nt; one may therefore set <code>floor=2</code> as a modest optimization for such data. YMMV</small></p>
<p>The indexing script will, among other steps, add a few <a href="https://sqlite.org/gencol.html">generated columns</a> to the original table. So if you later <code>SELECT * FROM tableName</code>, you'll get these extra values back (column names starting with <code>_gri_</code>). The extra columns are "virtual" so they don't take up space in the table itself, but they do end up populating the stored index.</p>
<p>At present, GRI cannot be used on <a href="https://www.sqlite.org/withoutrowid.html">WITHOUT ROWID</a> tables.</p>
<h3 id="query-gri">Query GRI</h3>
<p>The extension supplies a special SQL function to query a GRI-indexed table, generating the set of rowids identifying features that overlap a query range (queryChrom, queryBegin, queryEnd):</p>
<p><code>genomic_range_rowids(tableName, queryChrom, queryBegin, queryEnd[, ceiling, floor])</code></p>
<p>This is typically used to retrieve the result rows by selecting for <code>tableName._rowid_ IN genomic_range_rowids(...)</code>. For example,</p>
<div class="highlight"><pre><span></span><code><span class="k">SELECT</span> <span class="n">col1</span><span class="p">,</span> <span class="n">col2</span><span class="p">,</span> <span class="p">...</span> <span class="k">FROM</span> <span class="n">exons</span> <span class="k">WHERE</span> <span class="n">exons</span><span class="p">.</span><span class="n">_rowid_</span> <span class="k">IN</span>
  <span class="n">genomic_range_rowids</span><span class="p">(</span><span class="s1">&#39;exons&#39;</span><span class="p">,</span> <span class="s1">&#39;chr12&#39;</span><span class="p">,</span> <span class="mi">111803912</span><span class="p">,</span> <span class="mi">111804012</span><span class="p">)</span>
</code></pre></div>
<p>The queryChrom parameter might have SQL type TEXT or INTEGER, according to whether the GRI indexes name or rid.</p>
<p>The ordered rowid set identifies the features satisfying,</p>
<div class="highlight"><pre><span></span><code><span class="n">queryChrom</span> <span class="o">=</span> <span class="n">featureChrom</span> <span class="k">AND</span>
  <span class="k">NOT</span> <span class="p">(</span><span class="n">queryBegin</span> <span class="o">&gt;</span> <span class="n">featureEnd</span> <span class="k">OR</span> <span class="n">queryEnd</span> <span class="o">&lt;</span> <span class="n">featureBegin</span><span class="p">)</span>
</code></pre></div>
<p>(<em>"query is not disjoint from feature"</em>)</p>
<p><strong>❗ This includes features that <em>abut</em> as well as those that <em>overlap</em> the query range,</strong> per the half-open position convention. If you don't want those, or if you want only "contained" features, add a WHERE clause to your query (e.g. <code>WHERE _gri_beg &gt;= queryBeg AND _gri_beg+_gri_len &lt;= queryEnd</code>).</p>
<p><strong>❗ Results return in rowid order,</strong> which isn't necessarily genomic range order (see <em>Advice for big data</em>, below). Add an ORDER BY clause to your query if needed (e.g. <code>ORDER BY _gri_rid, _gri_beg, _gri_len</code>).</p>
<p><small>The query won't match any rows with NULL feature coordinates. If needed, the GRI can inform this query for NULL chromosome/rid: <code>SELECT ... FROM tableName WHERE _gri_rid IS NULL</code>.</small></p>
<h4 id="level-bounds-optimization">Level bounds optimization</h4>
<p>The optional, trailing <code>ceiling</code> &amp; <code>floor</code> arguments to <code>genomic_range_rowids()</code> optimize GRI queries by bounding their search <em>levels</em>, skipping steps that'd be useless in view of the overall length distribution of the indexed features. (See <a href="../internals/">Internals</a> for full explanation.)</p>
<p>The extension supplies a SQL helper function <code>genomic_range_index_levels(tableName)</code> to detect appropriate level bounds for the current version of the table. Example usage:</p>
<div class="highlight"><pre><span></span><code><span class="k">SELECT</span> <span class="n">col1</span><span class="p">,</span> <span class="n">col2</span><span class="p">,</span> <span class="p">...</span> <span class="k">FROM</span> <span class="n">exons</span><span class="p">,</span> <span class="n">genomic_range_index_levels</span><span class="p">(</span><span class="s1">&#39;exons&#39;</span><span class="p">)</span>
  <span class="k">WHERE</span> <span class="n">exons</span><span class="p">.</span><span class="n">_rowid_</span> <span class="k">IN</span>
    <span class="n">genomic_range_rowids</span><span class="p">(</span><span class="s1">&#39;exons&#39;</span><span class="p">,</span> <span class="s1">&#39;chr12&#39;</span><span class="p">,</span> <span class="mi">111803912</span><span class="p">,</span> <span class="mi">111804012</span><span class="p">,</span>
                         <span class="n">_gri_ceiling</span><span class="p">,</span> <span class="n">_gri_floor</span><span class="p">)</span>
</code></pre></div>
<p>Here <code>_gri_ceiling</code> and <code>_gri_floor</code> are columns of the single row computed by <code>genomic_range_index_levels('exons')</code>.</p>
<p><code>genomic_range_index_levels()</code> performs some upfront analysis of table's GRI upon its first use on any database connection. The cost of this analysis should be worthwhile if it's used to optimize many <code>genomic_range_rowids()</code> operations (but not just one or a few). Subsequent uses of <code>genomic_range_index_levels()</code> on the same connection &amp; table reuse the first analysis, unless the database changes in the meantime, in which case the analysis must be redone. This suggests using <code>genomic_range_index_levels()</code> only once the database is read-only.</p>
<p><small>Instead of detecting current bounds, they can be figured manually as follows. Set the integer ceiling to <em>C</em>, 0 &lt; <em>C</em> &lt; 16, such that all (present &amp; future) indexed features are guaranteed to have lengths &le;16<sup><em>C</em></sup>. For example, if you're querying features on the human genome, then you can set ceiling=7 because the lengthiest chromosome sequence is &lt;16<sup>7</sup>nt. Set the integer floor <em>F</em> to (i) the floor value supplied at GRI creation, if any; (ii) <em>F</em> &gt; 0 such that the minimum possible feature length &gt;16<sup><em>F</em>-1</sup>, if any; or (iii) zero. The safe, default bounds are C=15, F=0. GRI queries with inappropriate bounds are liable to produce incomplete results.</small></p>
<h4 id="joining-tables-on-range-overlap">Joining tables on range overlap</h4>
<p>Suppose we have two tables with genomic features to join on range overlap. Only the "right-hand" table must have a GRI; preferably the smaller of the two. For example, annotating a table of variants with the surrounding exon(s), if any:</p>
<div class="highlight"><pre><span></span><code><span class="k">SELECT</span> <span class="n">variants</span><span class="p">.</span><span class="o">*</span><span class="p">,</span> <span class="n">exons</span><span class="p">.</span><span class="n">_rowid_</span>
<span class="k">FROM</span> <span class="n">variants</span> <span class="k">LEFT</span> <span class="k">JOIN</span> <span class="n">exons</span> <span class="k">ON</span> <span class="n">exons</span><span class="p">.</span><span class="n">_rowid_</span> <span class="k">IN</span>
  <span class="n">genomic_range_rowids</span><span class="p">(</span>
    <span class="s1">&#39;exons&#39;</span><span class="p">,</span>
    <span class="n">variants</span><span class="p">.</span><span class="n">chrom</span><span class="p">,</span>
    <span class="n">variants</span><span class="p">.</span><span class="n">beginPos</span><span class="p">,</span>
    <span class="n">variants</span><span class="p">.</span><span class="n">endPos</span>
  <span class="p">)</span>
</code></pre></div>
<p>We fill out the GRI query range using the three coordinate columns of the variants table.</p>
<p>We may be able to speed this up by supplying level bounds, as discussed above:</p>
<div class="highlight"><pre><span></span><code><span class="k">SELECT</span> <span class="n">variants</span><span class="p">.</span><span class="o">*</span><span class="p">,</span> <span class="n">exons</span><span class="p">.</span><span class="n">_rowid_</span>
<span class="k">FROM</span> <span class="n">genomic_range_index_levels</span><span class="p">(</span><span class="s1">&#39;exons&#39;</span><span class="p">),</span>
     <span class="n">variants</span> <span class="k">LEFT</span> <span class="k">JOIN</span> <span class="n">exons</span> <span class="k">ON</span> <span class="n">exons</span><span class="p">.</span><span class="n">_rowid_</span> <span class="k">IN</span>
  <span class="n">genomic_range_rowids</span><span class="p">(</span>
    <span class="s1">&#39;exons&#39;</span><span class="p">,</span>
    <span class="n">variants</span><span class="p">.</span><span class="n">chrom</span><span class="p">,</span>
    <span class="n">variants</span><span class="p">.</span><span class="n">beginPos</span><span class="p">,</span>
    <span class="n">variants</span><span class="p">.</span><span class="n">endPos</span><span class="p">,</span>
    <span class="n">_gri_ceiling</span><span class="p">,</span>
    <span class="n">_gri_floor</span>
  <span class="p">)</span>
</code></pre></div>
<p>See also "Advice for big data" below on optimizing storage layout for GRI queries.</p>
<h3 id="reference-genome-metadata">Reference genome metadata</h3>
<p>The following routines support the aforementioned, recommended convention for storing a <code>_gri_refseq</code> table with information about the genomic reference sequences, which other tables can cross-reference by integer ID (rid) instead of storing textual chromosome names. The columns of <code>_gri_refseq</code> include:</p>
<ol>
<li><code>_gri_rid INTEGER PRIMARY KEY</code></li>
<li><code>gri_refseq_name TEXT NOT NULL</code></li>
<li><code>gri_refseq_length INTEGER NOT NULL</code></li>
<li><code>gri_assembly TEXT</code> genome assembly name (optional)</li>
<li><code>gri_refget_id TEXT</code> <a href="http://samtools.github.io/hts-specs/refget.html">refget</a> sequence ID (optional)</li>
<li><code>gri_refseq_meta_json TEXT DEFAULT '{}'</code> JSON object with arbitrary metadata</li>
</ol>
<p><strong>↪ Put Reference Assembly SQL:</strong> <em>Generate a string</em> containing a series of SQL statements which <em>when executed</em> creates <code>_gri_refseq</code> and populates it with information about a reference assembly whose details are bundled into the extension.</p>
<div class="tabbed-set" data-tabs="2:6"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><label for="__tabbed_2_1">Python</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="n">refseq_sql</span> <span class="o">=</span> <span class="n">genomicsqlite</span><span class="o">.</span><span class="n">put_reference_assembly_sql</span><span class="p">(</span>
  <span class="n">dbconn</span><span class="p">,</span> <span class="s1">&#39;GRCh38_no_alt_analysis_set&#39;</span>
<span class="p">)</span>
<span class="n">dbconn</span><span class="o">.</span><span class="n">executescript</span><span class="p">(</span><span class="n">refseq_sql</span><span class="p">)</span>
</code></pre></div>
</div>
<input id="__tabbed_2_2" name="__tabbed_2" type="radio" /><label for="__tabbed_2_2">Java</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="n">String</span> <span class="n">refSql</span> <span class="o">=</span> <span class="n">GenomicSQLite</span><span class="p">.</span><span class="na">putReferenceAssemblySQL</span><span class="p">(</span>
  <span class="n">dbconn</span><span class="p">,</span> <span class="s">&quot;GRCh38_no_alt_analysis_set&quot;</span>
<span class="p">);</span>
<span class="n">dbconn</span><span class="p">.</span><span class="na">createStatement</span><span class="p">().</span><span class="na">executeUpdate</span><span class="p">(</span><span class="n">refSql</span><span class="p">);</span>
</code></pre></div>
</div>
<input id="__tabbed_2_3" name="__tabbed_2" type="radio" /><label for="__tabbed_2_3">Rust</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="kd">let</span><span class="w"> </span><span class="n">ref_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dbconn</span><span class="w"></span>
<span class="w">  </span><span class="p">.</span><span class="n">put_reference_assembly_sql</span><span class="p">(</span><span class="s">&quot;GRCh38_no_alt_analysis_set&quot;</span><span class="p">)</span><span class="o">?</span><span class="p">;</span><span class="w"></span>
<span class="n">dbconn</span><span class="p">.</span><span class="n">execute_batch</span><span class="p">(</span><span class="o">&amp;</span><span class="n">ref_sql</span><span class="p">)</span><span class="o">?</span><span class="p">;</span><span class="w"></span>
</code></pre></div>
</div>
<input id="__tabbed_2_4" name="__tabbed_2" type="radio" /><label for="__tabbed_2_4">SQLiteCpp</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="nf">PutGenomicReferenceAssemblySQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">assembly</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">attached_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="c1">// SQLite::Database* dbconn in a transaction</span>
<span class="n">dbconn</span><span class="o">-&gt;</span><span class="n">exec</span><span class="p">(</span><span class="n">PutGenomicReferenceAssemblySQL</span><span class="p">(</span><span class="s">&quot;GRCh38_no_alt_analysis_set&quot;</span><span class="p">));</span><span class="w"></span>
</code></pre></div>
</div>
<input id="__tabbed_2_5" name="__tabbed_2" type="radio" /><label for="__tabbed_2_5">C++</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="nf">PutGenomicReferenceAssemblySQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">assembly</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">attached_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="n">refseq_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">PutGenomicReferenceAssemblySQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="s">&quot;GRCh38_no_alt_analysis_set&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>
<span class="c1">// sqlite3* dbconn in a transaction</span>
<span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">errmsg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">nullptr</span><span class="p">;</span><span class="w"></span>
<span class="kt">int</span><span class="w"> </span><span class="n">rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sqlite3_exec</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span><span class="w"> </span><span class="n">refseq_sql</span><span class="p">.</span><span class="n">c_str</span><span class="p">(),</span><span class="w"> </span><span class="k">nullptr</span><span class="p">,</span><span class="w"> </span><span class="k">nullptr</span><span class="p">,</span><span class="w"> </span><span class="o">&amp;</span><span class="n">errmsg</span><span class="p">);</span><span class="w"></span>
<span class="c1">// check rc, free errmsg</span>
</code></pre></div>
</div>
<input id="__tabbed_2_6" name="__tabbed_2" type="radio" /><label for="__tabbed_2_6">C</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="nf">put_genomic_reference_assembly_sql</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">assembly</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">attached_schema</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">refseq_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">put_genomic_reference_assembly_sql</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="s">&quot;GRCh38_no_alt_analysis_set&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">nullptr</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>
<span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">*</span><span class="n">refseq_sql</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">errmsg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">0</span><span class="p">;</span><span class="w"></span>
<span class="w">  </span><span class="cm">/* sqlite3* dbconn in a transaction */</span><span class="w"></span>
<span class="w">  </span><span class="kt">int</span><span class="w"> </span><span class="n">rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sqlite3_exec</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span><span class="w"> </span><span class="n">refseq_sql</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="o">&amp;</span><span class="n">errmsg</span><span class="p">);</span><span class="w"></span>
<span class="w">  </span><span class="cm">/* check rc, free errmsg */</span><span class="w"></span>
<span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="cm">/* see calling convention discussed in previous examples */</span><span class="w"></span>
<span class="p">}</span><span class="w"></span>
<span class="n">sqlite3_free</span><span class="p">(</span><span class="n">refseq_sql</span><span class="p">);</span><span class="w"></span>
</code></pre></div>
</div>
</div>
<p>Available assemblies:</p>
<ul>
<li><code>GRCh38_no_alt_analysis_set</code></li>
</ul>
<p><strong>↪ Put Reference Sequence SQL:</strong> <em>Generate a string</em> containing a series of SQL statements which <em>when executed</em> creates <code>_gri_refseq</code> (if it doesn't exist) and adds <em>one</em> reference sequence with supplied attributes.</p>
<div class="tabbed-set" data-tabs="3:6"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><label for="__tabbed_3_1">Python</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="n">refseq_sql</span> <span class="o">=</span> <span class="n">genomicsqlite</span><span class="o">.</span><span class="n">put_reference_sequence_sql</span><span class="p">(</span>
  <span class="n">dbconn</span><span class="p">,</span> <span class="s1">&#39;chr17&#39;</span><span class="p">,</span> <span class="mi">83257441</span>
  <span class="c1"># optional: assembly, refget_id, meta (dict), rid</span>
<span class="p">)</span>
<span class="n">dbconn</span><span class="o">.</span><span class="n">executescript</span><span class="p">(</span><span class="n">refseq_sql</span><span class="p">)</span>
</code></pre></div>
</div>
<input id="__tabbed_3_2" name="__tabbed_3" type="radio" /><label for="__tabbed_3_2">Java</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="n">String</span> <span class="n">refSql</span> <span class="o">=</span> <span class="n">GenomicSQLite</span><span class="p">.</span><span class="na">putReferenceSequenceSQL</span><span class="p">(</span>
  <span class="n">dbconn</span><span class="p">,</span> <span class="s">&quot;chr17&quot;</span><span class="p">,</span> <span class="mi">83257441L</span>
  <span class="c1">// optional overloads:</span>
  <span class="c1">// String assembly, String refget_id, String meta_json, long rid</span>
<span class="p">);</span>
<span class="n">dbconn</span><span class="p">.</span><span class="na">createStatement</span><span class="p">().</span><span class="na">executeUpdate</span><span class="p">(</span><span class="n">refSql</span><span class="p">);</span>
</code></pre></div>
</div>
<input id="__tabbed_3_3" name="__tabbed_3" type="radio" /><label for="__tabbed_3_3">Rust</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="kd">let</span><span class="w"> </span><span class="n">chr17</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">genomicsqlite</span>::<span class="n">RefSeq</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="n">rid</span>: <span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="w">           </span><span class="c1">// -1 = automatic</span>
<span class="w">  </span><span class="n">name</span>: <span class="s">&quot;chr17&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="n">length</span>: <span class="mi">83257441</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="n">assembly</span>: <span class="nb">None</span><span class="p">,</span><span class="w">    </span><span class="c1">// Option&lt;String&gt;</span>
<span class="w">  </span><span class="n">refget_id</span>: <span class="nb">None</span><span class="p">,</span><span class="w">   </span><span class="c1">// Option&lt;String&gt;</span>
<span class="w">  </span><span class="n">meta_json</span>: <span class="nc">json</span>::<span class="n">object</span>::<span class="n">Object</span>::<span class="n">new</span><span class="p">(),</span><span class="w">  </span><span class="c1">// meta_json</span>
<span class="p">};</span><span class="w"></span>
<span class="kd">let</span><span class="w"> </span><span class="n">ref_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dbconn</span><span class="p">.</span><span class="n">put_reference_sequence_sql</span><span class="p">(</span><span class="o">&amp;</span><span class="n">chr17</span><span class="p">)</span><span class="o">?</span><span class="p">;</span><span class="w"></span>
<span class="n">dbconn</span><span class="p">.</span><span class="n">execute_batch</span><span class="p">(</span><span class="o">&amp;</span><span class="n">ref_sql</span><span class="p">)</span><span class="o">?</span><span class="p">;</span><span class="w"></span>
</code></pre></div>
</div>
<input id="__tabbed_3_4" name="__tabbed_3" type="radio" /><label for="__tabbed_3_4">SQLiteCpp</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="nf">PutGenomicReferenceSequenceSQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">name</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="n">sqlite3_int64</span><span class="w"> </span><span class="n">length</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">assembly</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">refget_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">meta_json</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;{}&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="n">sqlite3_int64</span><span class="w"> </span><span class="n">rid</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">-1</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">attached_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="c1">// SQLite::Database* dbconn in a transaction</span>
<span class="n">dbconn</span><span class="o">-&gt;</span><span class="n">exec</span><span class="p">(</span><span class="n">PutGenomicReferenceSequenceSQL</span><span class="p">(</span><span class="s">&quot;chr17&quot;</span><span class="p">,</span><span class="w"> </span><span class="mi">83257441</span><span class="p">));</span><span class="w"></span>
</code></pre></div>
</div>
<input id="__tabbed_3_5" name="__tabbed_3" type="radio" /><label for="__tabbed_3_5">C++</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="nf">PutGenomicReferenceSequenceSQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">name</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="n">sqlite3_int64</span><span class="w"> </span><span class="n">length</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">assembly</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">refget_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">meta_json</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;{}&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="n">sqlite3_int64</span><span class="w"> </span><span class="n">rid</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">-1</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">attached_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="n">refseq_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">PutGenomicReferenceAssemblySQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="s">&quot;chr17&quot;</span><span class="p">,</span><span class="w"> </span><span class="mi">83257441</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>
<span class="c1">// sqlite3* dbconn in a transaction</span>
<span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">errmsg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">nullptr</span><span class="p">;</span><span class="w"></span>
<span class="kt">int</span><span class="w"> </span><span class="n">rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sqlite3_exec</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span><span class="w"> </span><span class="n">refseq_sql</span><span class="p">.</span><span class="n">c_str</span><span class="p">(),</span><span class="w"> </span><span class="k">nullptr</span><span class="p">,</span><span class="w"> </span><span class="k">nullptr</span><span class="p">,</span><span class="w"> </span><span class="o">&amp;</span><span class="n">errmsg</span><span class="p">);</span><span class="w"></span>
<span class="c1">// check rc, free errmsg</span>
</code></pre></div>
</div>
<input id="__tabbed_3_6" name="__tabbed_3" type="radio" /><label for="__tabbed_3_6">C</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="nf">put_genomic_reference_sequence_sql</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">name</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="n">sqlite3_int64</span><span class="w"> </span><span class="n">length</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">assembly</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">refget_id</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">meta_json</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="n">sqlite3_int64</span><span class="w"> </span><span class="n">rid</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">attached_schema</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">refseq_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">put_genomic_reference_sequence_sql</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="s">&quot;chr17&quot;</span><span class="p">,</span><span class="w"> </span><span class="mi">83257441</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="mi">-1</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>
<span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">*</span><span class="n">refseq_sql</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">errmsg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">0</span><span class="p">;</span><span class="w"></span>
<span class="w">  </span><span class="cm">/* sqlite3* dbconn in a transaction */</span><span class="w"></span>
<span class="w">  </span><span class="kt">int</span><span class="w"> </span><span class="n">rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sqlite3_exec</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span><span class="w"> </span><span class="n">refseq_sql</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="o">&amp;</span><span class="n">errmsg</span><span class="p">);</span><span class="w"></span>
<span class="w">  </span><span class="cm">/* check rc, free errmsg */</span><span class="w"></span>
<span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="cm">/* see calling convention discussed in previous examples */</span><span class="w"></span>
<span class="p">}</span><span class="w"></span>
<span class="n">sqlite3_free</span><span class="p">(</span><span class="n">refseq_sql</span><span class="p">);</span><span class="w"></span>
</code></pre></div>
</div>
</div>
<p>If the <code>rid</code> argument is omitted or -1 then it will be assigned automatically upon insertion.</p>
<p><strong>↪ Get Reference Sequences by Rid:</strong> create an in-memory lookup table of the previously-stored reference information, keyed by rid integer. Assumes the stored information is read-only by this point. This table is for the application code's convenience to read tables that use the rid convention. Such uses can be also be served by SQL join on the <code>_gri_refseq</code> table (see Cookbook).</p>
<div class="tabbed-set" data-tabs="4:6"><input checked="checked" id="__tabbed_4_1" name="__tabbed_4" type="radio" /><label for="__tabbed_4_1">Python</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">ReferenceSequence</span><span class="p">(</span><span class="n">NamedTuple</span><span class="p">):</span>
  <span class="n">rid</span><span class="p">:</span> <span class="nb">int</span>
  <span class="n">name</span><span class="p">:</span> <span class="nb">str</span>
  <span class="n">length</span><span class="p">:</span> <span class="nb">int</span>
  <span class="n">assembly</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
  <span class="n">refget_id</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
  <span class="n">meta</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span>

<span class="n">refseq_by_rid</span> <span class="o">=</span> <span class="n">genomicsqlite</span><span class="o">.</span><span class="n">get_reference_sequences_by_rid</span><span class="p">(</span><span class="n">dbconn</span><span class="p">)</span>
<span class="c1"># refseq_by_rid: Dict[int, ReferenceSequence]</span>
</code></pre></div>
</div>
<input id="__tabbed_4_2" name="__tabbed_4" type="radio" /><label for="__tabbed_4_2">Java</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">java.util.HashMap</span><span class="p">;</span>
<span class="kn">import</span> <span class="nn">net.mlin.genomicsqlite.ReferenceSequence</span><span class="p">;</span>
<span class="cm">/*</span>
<span class="cm">public class ReferenceSequence {</span>
<span class="cm">  public final long rid, length;</span>
<span class="cm">  public final String name, assembly, refgetId, metaJson;</span>
<span class="cm">}</span>
<span class="cm">*/</span>
<span class="n">HashMap</span><span class="o">&lt;</span><span class="n">Long</span><span class="p">,</span> <span class="n">ReferenceSequence</span><span class="o">&gt;</span> <span class="n">refseqByRid</span>
  <span class="o">=</span> <span class="n">GenomicSQLite</span><span class="p">.</span><span class="na">getReferenceSequencesByRid</span><span class="p">(</span><span class="n">dbconn</span><span class="p">);</span>
</code></pre></div>
</div>
<input id="__tabbed_4_3" name="__tabbed_4" type="radio" /><label for="__tabbed_4_3">Rust</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="cm">/*</span>
<span class="cm">struct RefSeq {</span>
<span class="cm">    rid: i64,</span>
<span class="cm">    name: String,</span>
<span class="cm">    length: i64,</span>
<span class="cm">    assembly: Option&lt;String&gt;,</span>
<span class="cm">    refget_id: Option&lt;String&gt;,</span>
<span class="cm">    meta_json: json::object::Object,</span>
<span class="cm">}</span>
<span class="cm">*/</span><span class="w"></span>
<span class="kd">let</span><span class="w"> </span><span class="n">refseqs</span>: <span class="nc">HashMap</span><span class="o">&lt;</span><span class="kt">i64</span><span class="p">,</span><span class="w"> </span><span class="n">genomicsqlite</span>::<span class="n">RefSeq</span><span class="o">&gt;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dbconn</span><span class="w"></span>
<span class="w">  </span><span class="p">.</span><span class="n">get_reference_sequences_by_rid</span><span class="p">()</span><span class="o">?</span><span class="p">;</span><span class="w"></span>
</code></pre></div>
</div>
<input id="__tabbed_4_4" name="__tabbed_4" type="radio" /><label for="__tabbed_4_4">SQLiteCpp</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="k">struct</span> <span class="nc">gri_refseq_t</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="kt">long</span><span class="w"> </span><span class="kt">long</span><span class="w"> </span><span class="n">rid</span><span class="p">,</span><span class="w"> </span><span class="n">length</span><span class="p">;</span><span class="w"></span>
<span class="w">  </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">assembly</span><span class="p">,</span><span class="w"> </span><span class="n">refget_id</span><span class="p">,</span><span class="w"> </span><span class="n">meta_json</span><span class="p">;</span><span class="w"></span>
<span class="p">};</span><span class="w"></span>
<span class="n">std</span><span class="o">::</span><span class="n">map</span><span class="o">&lt;</span><span class="kt">long</span><span class="w"> </span><span class="kt">long</span><span class="p">,</span><span class="w"> </span><span class="n">gri_refseq_t</span><span class="o">&gt;</span><span class="w"> </span><span class="n">GetGenomicReferenceSequencesByRid</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="n">sqlite3</span><span class="w"> </span><span class="o">*</span><span class="n">dbconn</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">assembly</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">attached_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="c1">// SQLite::Database* dbconn</span>
<span class="k">auto</span><span class="w"> </span><span class="n">refseq_by_rid</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GetGenomicReferenceSequencesByRid</span><span class="p">(</span><span class="n">dbconn</span><span class="o">-&gt;</span><span class="n">getHandle</span><span class="p">());</span><span class="w"></span>
</code></pre></div>
</div>
<input id="__tabbed_4_5" name="__tabbed_4" type="radio" /><label for="__tabbed_4_5">C++</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="k">struct</span> <span class="nc">gri_refseq_t</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="kt">long</span><span class="w"> </span><span class="kt">long</span><span class="w"> </span><span class="n">rid</span><span class="p">,</span><span class="w"> </span><span class="n">length</span><span class="p">;</span><span class="w"></span>
<span class="w">  </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">assembly</span><span class="p">,</span><span class="w"> </span><span class="n">refget_id</span><span class="p">,</span><span class="w"> </span><span class="n">meta_json</span><span class="p">;</span><span class="w"></span>
<span class="p">};</span><span class="w"></span>
<span class="n">std</span><span class="o">::</span><span class="n">map</span><span class="o">&lt;</span><span class="kt">long</span><span class="w"> </span><span class="kt">long</span><span class="p">,</span><span class="w"> </span><span class="n">gri_refseq_t</span><span class="o">&gt;</span><span class="w"> </span><span class="n">GetGenomicReferenceSequencesByRid</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="n">sqlite3</span><span class="w"> </span><span class="o">*</span><span class="n">dbconn</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">assembly</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">attached_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="c1">// sqlite3* dbconn</span>
<span class="k">auto</span><span class="w"> </span><span class="n">refseq_by_rid</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GetGenomicReferenceSequencesByRid</span><span class="p">(</span><span class="n">dbconn</span><span class="p">);</span><span class="w"></span>
</code></pre></div>
</div>
<input id="__tabbed_4_6" name="__tabbed_4" type="radio" /><label for="__tabbed_4_6">C</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="cm">/* Omitted for want of idiomatic map type; pull requests welcome! */</span><span class="w"></span>
</code></pre></div>
</div>
</div>
<p>The optional <code>assembly</code> argument restricts the retrieved sequences to those with matching <code>gri_assembly</code> value. However, mixing different assemblies in <code>_gri_refseq</code> is not recommended.</p>
<p><strong>↪ Get Reference Sequences by Name:</strong> create an in-memory lookup table of the previously-stored reference information, keyed by sequence name. Assumes the stored information is read-only by this point. This table is for the application code's convenience to translate name to rid whilst formulating queries or inserting features from a text source.</p>
<div class="tabbed-set" data-tabs="5:6"><input checked="checked" id="__tabbed_5_1" name="__tabbed_5" type="radio" /><label for="__tabbed_5_1">Python</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="k">class</span> <span class="nc">ReferenceSequence</span><span class="p">(</span><span class="n">NamedTuple</span><span class="p">):</span>
  <span class="n">rid</span><span class="p">:</span> <span class="nb">int</span>
  <span class="n">name</span><span class="p">:</span> <span class="nb">str</span>
  <span class="n">length</span><span class="p">:</span> <span class="nb">int</span>
  <span class="n">assembly</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
  <span class="n">refget_id</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
  <span class="n">meta</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span>

<span class="n">refseq_by_name</span> <span class="o">=</span> <span class="n">genomicsqlite</span><span class="o">.</span><span class="n">get_reference_sequences_by_name</span><span class="p">(</span><span class="n">dbconn</span><span class="p">)</span>
<span class="c1"># refseq_by_name: Dict[str, ReferenceSequence]</span>
</code></pre></div>
</div>
<input id="__tabbed_5_2" name="__tabbed_5" type="radio" /><label for="__tabbed_5_2">Java</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">java.util.HashMap</span><span class="p">;</span>
<span class="kn">import</span> <span class="nn">net.mlin.genomicsqlite.ReferenceSequence</span><span class="p">;</span>
<span class="cm">/*</span>
<span class="cm">public class ReferenceSequence {</span>
<span class="cm">  public final long rid, length;</span>
<span class="cm">  public final String name, assembly, refgetId, metaJson;</span>
<span class="cm">}</span>
<span class="cm">*/</span>
<span class="n">HashMap</span><span class="o">&lt;</span><span class="n">String</span><span class="p">,</span> <span class="n">ReferenceSequence</span><span class="o">&gt;</span> <span class="n">refseqByName</span>
  <span class="o">=</span> <span class="n">GenomicSQLite</span><span class="p">.</span><span class="na">getReferenceSequencesByName</span><span class="p">(</span><span class="n">dbconn</span><span class="p">);</span>
</code></pre></div>
</div>
<input id="__tabbed_5_3" name="__tabbed_5" type="radio" /><label for="__tabbed_5_3">Rust</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="cm">/*</span>
<span class="cm">struct RefSeq {</span>
<span class="cm">    rid: i64,</span>
<span class="cm">    name: String,</span>
<span class="cm">    length: i64,</span>
<span class="cm">    assembly: Option&lt;String&gt;,</span>
<span class="cm">    refget_id: Option&lt;String&gt;,</span>
<span class="cm">    meta_json: json::object::Object,</span>
<span class="cm">}</span>
<span class="cm">*/</span><span class="w"></span>
<span class="kd">let</span><span class="w"> </span><span class="n">refseqs</span>: <span class="nc">HashMap</span><span class="o">&lt;</span><span class="nb">String</span><span class="p">,</span><span class="w"> </span><span class="n">genomicsqlite</span>::<span class="n">RefSeq</span><span class="o">&gt;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dbconn</span><span class="w"></span>
<span class="w">  </span><span class="p">.</span><span class="n">get_reference_sequences_by_name</span><span class="p">()</span><span class="o">?</span><span class="p">;</span><span class="w"></span>
</code></pre></div>
</div>
<input id="__tabbed_5_4" name="__tabbed_5" type="radio" /><label for="__tabbed_5_4">SQLiteCpp</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="k">struct</span> <span class="nc">gri_refseq_t</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="kt">long</span><span class="w"> </span><span class="kt">long</span><span class="w"> </span><span class="n">rid</span><span class="p">,</span><span class="w"> </span><span class="n">length</span><span class="p">;</span><span class="w"></span>
<span class="w">  </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">assembly</span><span class="p">,</span><span class="w"> </span><span class="n">refget_id</span><span class="p">,</span><span class="w"> </span><span class="n">meta_json</span><span class="p">;</span><span class="w"></span>
<span class="p">};</span><span class="w"></span>
<span class="n">std</span><span class="o">::</span><span class="n">map</span><span class="o">&lt;</span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="p">,</span><span class="w"> </span><span class="n">gri_refseq_t</span><span class="o">&gt;</span><span class="w"> </span><span class="n">GetGenomicReferenceSequencesByName</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="n">sqlite3</span><span class="w"> </span><span class="o">*</span><span class="n">dbconn</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">assembly</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">attached_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="c1">// SQLite::Database* dbconn</span>
<span class="k">auto</span><span class="w"> </span><span class="n">refseq_by_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GetGenomicReferenceSequencesByName</span><span class="p">(</span><span class="n">dbconn</span><span class="o">-&gt;</span><span class="n">getHandle</span><span class="p">());</span><span class="w"></span>
</code></pre></div>
</div>
<input id="__tabbed_5_5" name="__tabbed_5" type="radio" /><label for="__tabbed_5_5">C++</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="k">struct</span> <span class="nc">gri_refseq_t</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="kt">long</span><span class="w"> </span><span class="kt">long</span><span class="w"> </span><span class="n">rid</span><span class="p">,</span><span class="w"> </span><span class="n">length</span><span class="p">;</span><span class="w"></span>
<span class="w">  </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">assembly</span><span class="p">,</span><span class="w"> </span><span class="n">refget_id</span><span class="p">,</span><span class="w"> </span><span class="n">meta_json</span><span class="p">;</span><span class="w"></span>
<span class="p">};</span><span class="w"></span>
<span class="n">std</span><span class="o">::</span><span class="n">map</span><span class="o">&lt;</span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="p">,</span><span class="w"> </span><span class="n">gri_refseq_t</span><span class="o">&gt;</span><span class="w"> </span><span class="n">GetGenomicReferenceSequencesByName</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="n">sqlite3</span><span class="w"> </span><span class="o">*</span><span class="n">dbconn</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">assembly</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">attached_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="c1">// sqlite3* dbconn</span>
<span class="k">auto</span><span class="w"> </span><span class="n">refseq_by_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GetGenomicReferenceSequencesByName</span><span class="p">(</span><span class="n">dbconn</span><span class="p">);</span><span class="w"></span>
</code></pre></div>
</div>
<input id="__tabbed_5_6" name="__tabbed_5" type="radio" /><label for="__tabbed_5_6">C</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code><span class="cm">/* Omitted for want of idiomatic map type; pull requests welcome! */</span><span class="w"></span>
</code></pre></div>
</div>
</div>
<h3 id="cookbook">Cookbook</h3>
<h4 id="rid-to-chromosome-name">rid to chromosome name</h4>
<p>Table identifies each feature's chromosome by rid, and we want to see them with text chromosome names.</p>
<div class="highlight"><pre><span></span><code><span class="k">SELECT</span> <span class="n">gri_refseq_name</span><span class="p">,</span> <span class="n">feature_table</span><span class="p">.</span><span class="o">*</span>
  <span class="k">FROM</span> <span class="n">feature_table</span> <span class="k">NATURAL</span> <span class="k">JOIN</span> <span class="n">_gri_refseq</span>
</code></pre></div>
<p>The join key here is <code>_gri_rid</code>, which is one of the generated columns added by GRI creation.</p>
<p>Alternatively, the application code can read rid from the row and translate it using the lookup table generated by the <strong>Get Reference Sequences by Rid</strong> routine.</p>
<h4 id="query-rid-using-chromosome-name">Query rid using chromosome name</h4>
<p>We're making a GRI query on a table that stores rid integers, but our query range has a chromosome name.</p>
<div class="highlight"><pre><span></span><code><span class="k">SELECT</span> <span class="n">feature_table</span><span class="p">.</span><span class="o">*</span> <span class="k">FROM</span>
  <span class="p">(</span><span class="k">SELECT</span> <span class="n">_gri_rid</span> <span class="k">AS</span> <span class="n">rid</span> <span class="k">FROM</span> <span class="n">_gri_refseq</span>
    <span class="k">WHERE</span> <span class="n">gri_refseq_name</span><span class="o">=</span><span class="s1">&#39;chr12&#39;</span><span class="p">)</span> <span class="k">AS</span> <span class="n">query</span><span class="p">,</span> <span class="n">feature_table</span>
  <span class="k">WHERE</span> <span class="n">feature_table</span><span class="p">.</span><span class="n">_rowid_</span> <span class="k">IN</span>
    <span class="n">genomic_range_rowids</span><span class="p">(</span><span class="s1">&#39;feature_table&#39;</span><span class="p">,</span><span class="n">query</span><span class="p">.</span><span class="n">rid</span><span class="p">,</span><span class="mi">111803912</span><span class="p">,</span><span class="mi">111804012</span><span class="p">)</span>
</code></pre></div>
<p>We use a subquery to look up the rid corresponding to the known chromosome name. Alternatively, the application code can first convert the query name to rid using the lookup table generated by the <strong>Get Reference Sequences by Name</strong> routine.</p>
<h4 id="circular-chromosome-query">Circular chromosome query</h4>
<p>On circular chromosomes, range queries should include features that wrap around the origin to end inside the desired range. If we've stored them naively with featureEnd = featureBegin + featureLength, then we can build a unified query with reference to the stored chromosome lengths:</p>
<div class="highlight"><pre><span></span><code><span class="k">SELECT</span> <span class="n">col1</span><span class="p">,</span> <span class="n">col2</span><span class="p">,</span> <span class="p">...</span> <span class="k">FROM</span>
  <span class="p">(</span><span class="k">SELECT</span> <span class="n">gri_refseq_length</span> <span class="k">FROM</span> <span class="n">_gri_refseq</span> <span class="k">WHERE</span> <span class="n">_gri_rid</span> <span class="o">=</span> <span class="n">queryRid</span><span class="p">),</span>
  <span class="n">featureTable</span> <span class="k">WHERE</span> <span class="n">featureTable</span><span class="p">.</span><span class="n">_rowid_</span> <span class="k">IN</span>
  <span class="p">(</span><span class="n">genomic_range_rowids</span><span class="p">(</span><span class="s1">&#39;featureTable&#39;</span><span class="p">,</span> <span class="n">queryRid</span><span class="p">,</span> <span class="n">queryBegin</span><span class="p">,</span> <span class="n">queryEnd</span><span class="p">)</span>
   <span class="k">UNION</span>
   <span class="n">genomic_range_rowids</span><span class="p">(</span>
     <span class="s1">&#39;featureTable&#39;</span><span class="p">,</span>
     <span class="n">queryRid</span><span class="p">,</span>
     <span class="n">gri_refseq_length</span><span class="o">+</span><span class="n">queryBegin</span><span class="p">,</span>
     <span class="n">gri_refseq_length</span><span class="o">+</span><span class="n">queryEnd</span><span class="p">))</span>
</code></pre></div>
<p>We query a second range beyond the chromosome length, which will match features that wrap around into the query. <code>UNION</code> deduplicates the result rowids.</p>
<p>As a convention, set <code>"circular": true</code> in the <code>_gri_refseq.gri_refseq_meta_json</code> for circular chromosomes.</p>
<h3 id="advice-for-big-data">Advice for big data</h3>
<p>The database file stores tables in rowid order (effectively). It's therefore preferable for a mainly-GRI-queried table to be written in genomic range order, so that the features' (chromosome, beginPosition) monotonically increase with rowid, and range queries enjoy storage/cache locality. See <a href="../guide_db/#optimizing-storage-layout"><em>Optimizing storage layout</em></a> in the compression guide for advice if it isn't straightforward to initally write the rows ordered by (chromosome, beginPosition). Though not required in theory, this may be needed in practice for GRI queries that will match a material fraction of a big table's rows.</p>
<p>A series of many GRI queries (including in service of a join) should also proceed in genomic range order. If this isn't possible, then ideally the database page cache should be enlarged to fit the entire indexed table in memory.</p>
<p>If you expect a GRI query to yield a very large, contiguous rowid result set (e.g. all features on a chromosome, in a table <em>known</em> to be range-sorted), then the following specialized query plan may be advantageous:</p>
<ol>
<li>Ask GRI for <em>first</em> relevant rowid, <code>SELECT MIN(_rowid_) AS firstRowid FROM genomic_range_rowids(...)</code></li>
<li>Open a cursor on <code>SELECT ... FROM tableName WHERE _rowid_ &gt;= firstRowid</code></li>
<li>Loop through rows for as long as they're relevant.</li>
</ol>
<p>But this plan strongly depends on the contiguity assumption.</p>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../guide_db/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Opening Compressed Databases" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Opening Compressed Databases
            </div>
          </div>
        </a>
      
      
        
        <a href="../guide_helpers/" class="md-footer__link md-footer__link--next" aria-label="Next: Useful Routines" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Useful Routines
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        
          Made with
          <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
            Material for MkDocs
          </a>
        
        
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../assets/javascripts/workers/search.fcfe8b6d.min.js", "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.b1047164.min.js"></script>
      
    
  </body>
</html>