
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.3.1, mkdocs-material-8.4.0">
    
    
      
        <title>Useful Routines - Genomics Extension for SQLite</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.69437709.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.cbb835fc.min.css">
        
      
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#programming-guide-useful-routines" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Genomics Extension for SQLite" class="md-header__button md-logo" aria-label="Genomics Extension for SQLite" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Genomics Extension for SQLite
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Useful Routines
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Genomics Extension for SQLite" class="md-nav__button md-logo" aria-label="Genomics Extension for SQLite" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Genomics Extension for SQLite
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Intro & Install
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../guide_db/" class="md-nav__link">
        Opening Compressed Databases
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../guide_gri/" class="md-nav__link">
        Genomic Range Indexing
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Useful Routines
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Useful Routines
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#dna-reverse-complement" class="md-nav__link">
    DNA reverse complement
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parse-genomic-range-text" class="md-nav__link">
    Parse genomic range text
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#two-bit-encoding-for-nucleotide-sequences" class="md-nav__link">
    Two-bit encoding for nucleotide sequences
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#json1-and-uint-extensions" class="md-nav__link">
    JSON1 and UINT extensions
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#attach-genomicsqlite-database" class="md-nav__link">
    Attach GenomicSQLite database
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#compress-existing-sqlite3-database" class="md-nav__link">
    Compress existing SQLite3 database
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#genomics-extension-version" class="md-nav__link">
    Genomics Extension version
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../bindings/" class="md-nav__link">
        Writing Language Bindings
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../internals/" class="md-nav__link">
        Internals
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#dna-reverse-complement" class="md-nav__link">
    DNA reverse complement
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parse-genomic-range-text" class="md-nav__link">
    Parse genomic range text
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#two-bit-encoding-for-nucleotide-sequences" class="md-nav__link">
    Two-bit encoding for nucleotide sequences
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#json1-and-uint-extensions" class="md-nav__link">
    JSON1 and UINT extensions
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#attach-genomicsqlite-database" class="md-nav__link">
    Attach GenomicSQLite database
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#compress-existing-sqlite3-database" class="md-nav__link">
    Compress existing SQLite3 database
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#genomics-extension-version" class="md-nav__link">
    Genomics Extension version
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                


<h1 id="programming-guide-useful-routines">Programming Guide - Useful Routines</h1>
<h3 id="dna-reverse-complement">DNA reverse complement</h3>
<p>Reverse-complements a DNA text value (containing only characters from the set <code>AGCTagct</code>), preserving original case.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="1:1"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">SQL</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="k">SELECT</span><span class="w"> </span><span class="n">dna_revcomp</span><span class="p">(</span><span class="s1">&#39;AGCTagct&#39;</span><span class="p">)</span><span class="w">  </span><span class="c1">-- &#39;agctAGCT&#39;</span>
</code></pre></div>
</div>
</div>
</div>
<p>Given NULL, returns NULL. Any other input is an error.</p>
<h3 id="parse-genomic-range-text">Parse genomic range text</h3>
<p>These SQL functions process a text value like <code>'chr1:2,345-6,789'</code> into its three parts (sequence/chromosome name, begin position, and end position).</p>
<div class="tabbed-set tabbed-alternate" data-tabs="2:1"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><div class="tabbed-labels"><label for="__tabbed_2_1">SQL</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="k">SELECT</span><span class="w"> </span><span class="n">parse_genomic_range_sequence</span><span class="p">(</span><span class="s1">&#39;chr1:2,345-6,789&#39;</span><span class="p">)</span><span class="w">  </span><span class="c1">-- &#39;chr1&#39;</span>
<span class="k">SELECT</span><span class="w"> </span><span class="n">parse_genomic_range_begin</span><span class="p">(</span><span class="s1">&#39;chr1:2,345-6,789&#39;</span><span class="p">)</span><span class="w">     </span><span class="c1">-- 2344 (!)</span>
<span class="k">SELECT</span><span class="w"> </span><span class="n">parse_genomic_range_end</span><span class="p">(</span><span class="s1">&#39;chr1:2,345-6,789&#39;</span><span class="p">)</span><span class="w">       </span><span class="c1">-- 6789</span>
</code></pre></div>
</div>
</div>
</div>
<p>❗ Since such text ranges are conventionally one-based and closed, <code>parse_genomic_range_begin()</code> effectively converts them to zero-based and half-open by <a href="https://genome.ucsc.edu/FAQ/FAQtracks#tracks1">returning one less than the text begin position</a>.</p>
<p>Given NULL, each function returns NULL. An error is raised if the text value can't be parsed, or for any other input type.</p>
<h3 id="two-bit-encoding-for-nucleotide-sequences">Two-bit encoding for nucleotide sequences</h3>
<p>The extension supplies SQL functions to pack a DNA/RNA sequence TEXT value into a smaller BLOB value, using two bits per nucleotide. (Review <a href="https://www.sqlite.org/datatype3.html">SQLite Datatypes</a> on the important differences between TEXT and BLOB values &amp; columns.) Storing a large database of sequences using such BLOBs instead of TEXT can improve application I/O efficiency, with up to 4X more nucleotides cached in the same memory space. It is not, however, expected to greatly shrink the database file on disk, owing to the automatic storage compression.</p>
<p>The encoding is case-insensitive and considers <code>T</code> and <code>U</code> equivalent.</p>
<p><em>Encoding:</em></p>
<div class="tabbed-set tabbed-alternate" data-tabs="3:1"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><div class="tabbed-labels"><label for="__tabbed_3_1">SQL</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="k">SELECT</span><span class="w"> </span><span class="n">nucleotides_twobit</span><span class="p">(</span><span class="s1">&#39;TCAG&#39;</span><span class="p">)</span><span class="w"></span>
</code></pre></div>
</div>
</div>
</div>
<p>Given a TEXT value consisting of characters from the set <code>ACGTUacgtu</code>, compute a two-bit-encoded BLOB value that can later be decoded using <code>twobit_dna()</code> or <code>twobit_rna()</code>. Given any other ASCII TEXT value (including empty), pass it through unchanged as TEXT. Given NULL, return NULL. Any other input is an error.</p>
<p>Typically used to populate a BLOB column <code>C</code> with e.g.</p>
<div class="highlight"><pre><span></span><code><span class="k">INSERT</span><span class="w"> </span><span class="k">INTO</span><span class="w"> </span><span class="n">some_table</span><span class="p">(...,</span><span class="k">C</span><span class="p">)</span><span class="w"> </span><span class="k">VALUES</span><span class="p">(...,</span><span class="n">nucleotides_twobit</span><span class="p">(</span><span class="o">?</span><span class="p">))</span><span class="w"></span>
</code></pre></div>
<p>This works even if some of the sequences contain <code>N</code>s or other characters, in which case those sequences are stored as the original TEXT values. Make sure the column has schema type <code>BLOB</code> to avoid spurious coercions, and by convention, the column should be named *_twobit.</p>
<p><em>Decoding:</em></p>
<div class="tabbed-set tabbed-alternate" data-tabs="4:1"><input checked="checked" id="__tabbed_4_1" name="__tabbed_4" type="radio" /><div class="tabbed-labels"><label for="__tabbed_4_1">SQL</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="k">SELECT</span><span class="w"> </span><span class="n">twobit_dna</span><span class="p">(</span><span class="n">nucleotides_twobit</span><span class="p">(</span><span class="s1">&#39;TCAG&#39;</span><span class="p">))</span><span class="w"></span>
<span class="k">SELECT</span><span class="w"> </span><span class="n">twobit_rna</span><span class="p">(</span><span class="n">nucleotides_twobit</span><span class="p">(</span><span class="s1">&#39;UCAG&#39;</span><span class="p">))</span><span class="w"></span>
<span class="k">SELECT</span><span class="w"> </span><span class="n">twobit_dna</span><span class="p">(</span><span class="n">nucleotides_twobit</span><span class="p">(</span><span class="s1">&#39;TCAG&#39;</span><span class="p">),</span><span class="n">Y</span><span class="p">,</span><span class="n">Z</span><span class="p">)</span><span class="w"></span>
<span class="k">SELECT</span><span class="w"> </span><span class="n">twobit_rna</span><span class="p">(</span><span class="n">nucleotides_twobit</span><span class="p">(</span><span class="s1">&#39;UCAG&#39;</span><span class="p">),</span><span class="n">Y</span><span class="p">,</span><span class="n">Z</span><span class="p">)</span><span class="w"></span>
</code></pre></div>
</div>
</div>
</div>
<p>Given a two-bit-encoded BLOB value, decode the nucleotide sequence as uppercased TEXT, with <code>T</code>'s for <code>twobit_dna()</code> and <code>U</code>'s for <code>twobit_rna()</code>. Given a TEXT value, pass it through unchanged. Given NULL, return NULL. Any other first input is an error.</p>
<p>The optional <code>Y</code> and <code>Z</code> arguments can be used to compute <a href="https://sqlite.org/lang_corefunc.html#substr"><code>substr(twobit_dna(X),Y,Z)</code></a> more efficiently, without decoding the whole sequence. <small>Unfortunately however, <a href="https://sqlite.org/forum/forumpost/756c1a1e48?t=h">SQLite internals</a> make this operation still liable to use time &amp; memory proportional to the full length of X, not Z. If frequent random access into long sequences is needed, then consider splitting them across multiple rows.</small></p>
<p>Take care to only use BLOBs originally produced by <code>nucleotides_twobit()</code>, as other BLOBs may decode to spurious nucleotide sequences. If you <code>SELECT twobit_dna(C) FROM some_table</code> on a column with mixed BLOB and TEXT values as suggested above, note that the results actually stored as TEXT preserve their case and T/U letters, unlike decoded BLOBs.</p>
<p><em>Length:</em></p>
<div class="tabbed-set tabbed-alternate" data-tabs="5:1"><input checked="checked" id="__tabbed_5_1" name="__tabbed_5" type="radio" /><div class="tabbed-labels"><label for="__tabbed_5_1">SQL</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="k">SELECT</span><span class="w"> </span><span class="n">twobit_length</span><span class="p">(</span><span class="n">dna_twobit</span><span class="p">(</span><span class="s1">&#39;TCAG&#39;</span><span class="p">))</span><span class="w"></span>
</code></pre></div>
</div>
</div>
</div>
<p>Given a two-bit-encoded BLOB value, return the length of the <em>decoded</em> sequence (without actually decoding it). This is <em>not</em> equal to <code>4*length(BLOB)</code> due to padding.</p>
<p>Given a TEXT value, return its byte length. Given NULL, return NULL. Any other input is an error.</p>
<h3 id="json1-and-uint-extensions">JSON1 and UINT extensions</h3>
<p>The Genomics Extension bundles the SQLite developers' <a href="https://www.sqlite.org/json1.html">JSON1 extension</a> and enables it automatically. By convention, JSON object columns should be named *_json and JSON array columns should be named *_jsarray. The JSON1 functions can be used with <a href="https://sqlite.org/gencol.html">generated columns</a> to effectively allow indexing of JSON-embedded fields.</p>
<p>The <a href="https://www.sqlite.org/uintcseq.html">UINT collating sequence</a> is also bundled. This can be useful to make e.g. <code>ORDER BY chromosome COLLATE UINT</code> put 'chr2' before 'chr10'.</p>
<h3 id="attach-genomicsqlite-database">Attach GenomicSQLite database</h3>
<p><strong>↪ GenomicSQLite Attach:</strong> <em>Generate a string</em> containing a series of SQL statements to execute on an existing database connection in order to <a href="https://www.sqlite.org/lang_attach.html">ATTACH</a> a GenomicSQLite database under a given schema name. The main connection may be a plain, uncompressed SQLite3 database, as long as (i) it was opened with the <code>SQLITE_OPEN_URI</code> flag or language equivalent and (ii) the Genomics Extension is loaded in the executing program.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="6:6"><input checked="checked" id="__tabbed_6_1" name="__tabbed_6" type="radio" /><input id="__tabbed_6_2" name="__tabbed_6" type="radio" /><input id="__tabbed_6_3" name="__tabbed_6" type="radio" /><input id="__tabbed_6_4" name="__tabbed_6" type="radio" /><input id="__tabbed_6_5" name="__tabbed_6" type="radio" /><input id="__tabbed_6_6" name="__tabbed_6" type="radio" /><div class="tabbed-labels"><label for="__tabbed_6_1">Python</label><label for="__tabbed_6_2">Java</label><label for="__tabbed_6_3">Rust</label><label for="__tabbed_6_4">SQLiteCpp</label><label for="__tabbed_6_5">C++</label><label for="__tabbed_6_6">C</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">dbconn</span> <span class="o">=</span> <span class="n">sqlite3</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="s1">&#39;any.db&#39;</span><span class="p">,</span> <span class="n">uri</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">attach_sql</span> <span class="o">=</span> <span class="n">genomicsqlite</span><span class="o">.</span><span class="n">attach_sql</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span> <span class="s1">&#39;compressed.db&#39;</span><span class="p">,</span> <span class="s1">&#39;db2&#39;</span><span class="p">)</span>
<span class="c1"># attach_sql() also takes configuration keyword arguments like</span>
<span class="c1"># genomicsqlite.connect()</span>
<span class="n">dbconn</span><span class="o">.</span><span class="n">executescript</span><span class="p">(</span><span class="n">attach_sql</span><span class="p">)</span>
<span class="c1"># compressed.db now attached as db2</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="c1">// If needed, no-op to trigger initial load of Genomics Extension:</span><span class="w"></span>
<span class="n">DriverManager</span><span class="p">.</span><span class="na">getConnection</span><span class="p">(</span><span class="s">&quot;jdbc:genomicsqlite::memory:&quot;</span><span class="p">);</span><span class="w"></span>

<span class="n">Connection</span><span class="w"> </span><span class="n">dbconn</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">DriverManager</span><span class="p">.</span><span class="na">getConnection</span><span class="p">(</span><span class="s">&quot;jdbc:sqlite:any.db&quot;</span><span class="p">);</span><span class="w"></span>
<span class="n">String</span><span class="w"> </span><span class="n">attachSql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GenomicSQLite</span><span class="p">.</span><span class="na">attachSQL</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;compressed.db&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;db2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;{}&quot;</span><span class="p">);</span><span class="w"></span>
<span class="c1">//                                                             config_json ^^^^</span><span class="w"></span>
<span class="n">dbconn</span><span class="p">.</span><span class="na">createStatement</span><span class="p">().</span><span class="na">executeUpdate</span><span class="p">(</span><span class="n">attachSql</span><span class="p">);</span><span class="w"></span>
<span class="c1">// compressed.db now attached as db2</span><span class="w"></span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="kd">let</span><span class="w"> </span><span class="n">genomicsqlite_options</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">json</span>::<span class="n">object</span>::<span class="n">Object</span>::<span class="n">new</span><span class="p">();</span><span class="w"></span>
<span class="c1">// If needed, trigger initial load of Genomics Extension prior to</span>
<span class="c1">// opening other connections</span>
<span class="kd">let</span><span class="w"> </span><span class="n">nop</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">genomicsqlite</span>::<span class="n">open</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="s">&quot;:memory:&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="n">OpenFlags</span>::<span class="n">SQLITE_OPEN_CREATE</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">OpenFlags</span>::<span class="n">SQLITE_OPEN_READ_WRITE</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="o">&amp;</span><span class="n">genomicsqlite_options</span><span class="w"></span>
<span class="p">)</span><span class="o">?</span><span class="p">;</span><span class="w"></span>

<span class="c1">// dbconn: rusqlite::Connection with SQLITE_OPEN_URI</span>
<span class="kd">let</span><span class="w"> </span><span class="n">attach_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dbconn</span><span class="w"></span>
<span class="w">  </span><span class="p">.</span><span class="n">genomicsqlite_attach_sql</span><span class="p">(</span><span class="s">&quot;compressed.db&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;db2&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">                            </span><span class="o">&amp;</span><span class="n">genomicsqlite_options</span><span class="p">)</span><span class="o">?</span><span class="p">;</span><span class="w"></span>
<span class="n">dbconn</span><span class="p">.</span><span class="n">execute_batch</span><span class="p">(</span><span class="o">&amp;</span><span class="n">attach_sql</span><span class="p">)</span><span class="o">?</span><span class="p">;</span><span class="w"></span>
<span class="c1">// compressed.db is now attached as db2</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="nf">GenomicSQLiteAttachSQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">dbfile</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">schema_name</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">config_json</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;{}&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="n">attach_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GenomicSQLiteAttachSQL</span><span class="p">(</span><span class="s">&quot;compressed.db&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;db2&quot;</span><span class="p">);</span><span class="w"></span>
<span class="n">SQLite</span><span class="o">::</span><span class="n">Database</span><span class="w"> </span><span class="nf">dbconn</span><span class="p">(</span><span class="s">&quot;any.db&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">SQLITE_OPEN_URI</span><span class="p">);</span><span class="w"></span>
<span class="n">dbconn</span><span class="p">.</span><span class="n">exec</span><span class="p">(</span><span class="n">attach_sql</span><span class="p">);</span><span class="w"></span>
<span class="c1">// compressed.db now attached as db2</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="nf">GenomicSQLiteAttachSQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">dbfile</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">schema_name</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">config_json</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;{}&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="n">attach_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GenomicSQLiteAttachSQL</span><span class="p">(</span><span class="s">&quot;compressed.db&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;db2&quot;</span><span class="p">);</span><span class="w"></span>
<span class="c1">// sqlite3* dbconn opened using sqlite3_open_v2() on some db</span>
<span class="c1">//   with SQLITE_OPEN_URI</span>
<span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">errmsg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">nullptr</span><span class="p">;</span><span class="w"></span>
<span class="kt">int</span><span class="w"> </span><span class="n">rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sqlite3_exec</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span><span class="w"> </span><span class="n">attach_sql</span><span class="p">.</span><span class="n">c_str</span><span class="p">(),</span><span class="w"> </span><span class="k">nullptr</span><span class="p">,</span><span class="w"> </span><span class="k">nullptr</span><span class="p">,</span><span class="w"> </span><span class="o">&amp;</span><span class="n">errmsg</span><span class="p">);</span><span class="w"></span>
<span class="c1">// check rc, free errmsg</span>

<span class="c1">// compressed.db now attached as db2</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="nf">genomicsqlite_attach_sql</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">dbfile</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">schema_name</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">config_json</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">attach_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">genomicsqlite_attach_sql</span><span class="p">(</span><span class="s">&quot;compressed.db&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;db2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;{}&quot;</span><span class="p">);</span><span class="w"></span>
<span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">*</span><span class="n">attach_sql</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">errmsg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">0</span><span class="p">;</span><span class="w"></span>
<span class="w">  </span><span class="cm">/* sqlite3* dbconn opened using sqlite3_open_v2() on some db</span>
<span class="cm">   * with SQLITE_OPEN_URI */</span><span class="w"></span>
<span class="w">  </span><span class="kt">int</span><span class="w"> </span><span class="n">rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sqlite3_exec</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span><span class="w"> </span><span class="n">attach_sql</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="o">&amp;</span><span class="n">errmsg</span><span class="p">);</span><span class="w"></span>
<span class="w">  </span><span class="cm">/* check rc, free errmsg */</span><span class="w"></span>
<span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w"> </span><span class="cm">/* see calling convention discussed in previous examples */</span><span class="w"></span>
<span class="p">}</span><span class="w"></span>
<span class="n">sqlite3_free</span><span class="p">(</span><span class="n">attach_sql</span><span class="p">);</span><span class="w"></span>

<span class="cm">/* compressed.db now attached as db2 */</span><span class="w"></span>
</code></pre></div>
</div>
</div>
</div>
<p>❗ The file and schema names are textually pasted into a template SQL script. Take care to prevent SQL injection, if they're in any way determined by external input.</p>
<h3 id="compress-existing-sqlite3-database">Compress existing SQLite3 database</h3>
<p><strong>↪ GenomicSQLite Vacuum Into:</strong> <em>Generate a string</em> containing a series of SQL statements to execute on an existing database in order to copy it into a new compressed &amp; <a href="https://www.sqlite.org/lang_vacuum.html">defragmented</a> file. The source database may be a plain, uncompressed SQLite3 database, as long as (i) it was opened with the <code>SQLITE_OPEN_URI</code> flag or language equivalent and (ii) the Genomics Extension is loaded in the executing program.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="7:6"><input checked="checked" id="__tabbed_7_1" name="__tabbed_7" type="radio" /><input id="__tabbed_7_2" name="__tabbed_7" type="radio" /><input id="__tabbed_7_3" name="__tabbed_7" type="radio" /><input id="__tabbed_7_4" name="__tabbed_7" type="radio" /><input id="__tabbed_7_5" name="__tabbed_7" type="radio" /><input id="__tabbed_7_6" name="__tabbed_7" type="radio" /><div class="tabbed-labels"><label for="__tabbed_7_1">Python</label><label for="__tabbed_7_2">Java</label><label for="__tabbed_7_3">Rust</label><label for="__tabbed_7_4">SQLiteCpp</label><label for="__tabbed_7_5">C++</label><label for="__tabbed_7_6">C</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">dbconn</span> <span class="o">=</span> <span class="n">sqlite3</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="s1">&#39;existing.db&#39;</span><span class="p">,</span> <span class="n">uri</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">vacuum_sql</span> <span class="o">=</span> <span class="n">genomicsqlite</span><span class="o">.</span><span class="n">vacuum_into_sql</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span> <span class="s1">&#39;compressed.db&#39;</span><span class="p">)</span>
<span class="c1"># vacuum_into_sql() also takes configuration keyword arguments like</span>
<span class="c1"># genomicsqlite.connect() to control compression level &amp; page sizes</span>

<span class="n">dbconn</span><span class="o">.</span><span class="n">executescript</span><span class="p">(</span><span class="n">vacuum_sql</span><span class="p">)</span>
<span class="n">dbconn2</span> <span class="o">=</span> <span class="n">genomicsqlite</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="s1">&#39;compressed.db&#39;</span><span class="p">)</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="c1">// If needed, no-op to trigger initial load of Genomics Extension:</span><span class="w"></span>
<span class="n">DriverManager</span><span class="p">.</span><span class="na">getConnection</span><span class="p">(</span><span class="s">&quot;jdbc:genomicsqlite::memory:&quot;</span><span class="p">);</span><span class="w"></span>

<span class="n">Connection</span><span class="w"> </span><span class="n">dbconn</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">DriverManager</span><span class="p">.</span><span class="na">getConnection</span><span class="p">(</span><span class="s">&quot;jdbc:sqlite:existing.db&quot;</span><span class="p">);</span><span class="w"></span>
<span class="n">String</span><span class="w"> </span><span class="n">vacuumSql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GenomicSQLite</span><span class="p">.</span><span class="na">vacuumIntoSQL</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;compressed.db&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;{}&quot;</span><span class="p">);</span><span class="w"></span>
<span class="c1">//                                                          config_json ^^^^</span><span class="w"></span>
<span class="n">dbconn</span><span class="p">.</span><span class="na">createStatement</span><span class="p">().</span><span class="na">executeUpdate</span><span class="p">(</span><span class="n">vaccumSql</span><span class="p">);</span><span class="w"></span>
<span class="n">Connection</span><span class="w"> </span><span class="n">dbconn2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">DriverManager</span><span class="p">.</span><span class="na">getConnection</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="s">&quot;jdbc:genomicsqlite:compressed.db&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="kd">let</span><span class="w"> </span><span class="n">genomicsqlite_options</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">json</span>::<span class="n">object</span>::<span class="n">Object</span>::<span class="n">new</span><span class="p">();</span><span class="w"></span>
<span class="c1">// If needed, trigger initial load of Genomics Extension prior to</span>
<span class="c1">// opening other connections</span>
<span class="kd">let</span><span class="w"> </span><span class="n">nop</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">genomicsqlite</span>::<span class="n">open</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="s">&quot;:memory:&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="n">OpenFlags</span>::<span class="n">SQLITE_OPEN_CREATE</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">OpenFlags</span>::<span class="n">SQLITE_OPEN_READ_WRITE</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="o">&amp;</span><span class="n">genomicsqlite_options</span><span class="w"></span>
<span class="p">)</span><span class="o">?</span><span class="p">;</span><span class="w"></span>

<span class="c1">// dbconn: rusqlite::Connection with SQLITE_OPEN_URI</span>
<span class="kd">let</span><span class="w"> </span><span class="n">vacuum_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dbconn</span><span class="w"></span>
<span class="w">  </span><span class="p">.</span><span class="n">genomicsqlite_vacuum_into_sql</span><span class="p">(</span><span class="s">&quot;compressed.db&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">                                 </span><span class="o">&amp;</span><span class="n">genomicsqlite_options</span><span class="p">)</span><span class="o">?</span><span class="p">;</span><span class="w"></span>
<span class="n">dbconn</span><span class="p">.</span><span class="n">execute_batch</span><span class="p">(</span><span class="o">&amp;</span><span class="n">vacuum_sql</span><span class="p">)</span><span class="o">?</span><span class="p">;</span><span class="w"></span>
<span class="kd">let</span><span class="w"> </span><span class="n">dbconn2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">genomicsqlite</span>::<span class="n">open</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="s">&quot;compressed.db&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="n">OpenFlags</span>::<span class="n">SQLITE_OPEN_READ_WRITE</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="o">&amp;</span><span class="n">genomicsqlite_options</span><span class="w"></span>
<span class="p">)</span><span class="o">?</span><span class="p">;</span><span class="w"></span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="nf">GenomicSQLiteVacuumIntoSQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">dest_filename</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">config_json</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;{}&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="n">vacuum_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GenomicSQLiteVacuumIntoSQL</span><span class="p">(</span><span class="s">&quot;compressed.db&quot;</span><span class="p">);</span><span class="w"></span>
<span class="n">SQLite</span><span class="o">::</span><span class="n">Database</span><span class="w"> </span><span class="nf">dbconn</span><span class="p">(</span><span class="s">&quot;existing.db&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">SQLITE_OPEN_READONLY</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">SQLITE_OPEN_URI</span><span class="p">);</span><span class="w"></span>
<span class="n">dbconn</span><span class="p">.</span><span class="n">exec</span><span class="p">(</span><span class="n">vacuum_sql</span><span class="p">);</span><span class="w"></span>
<span class="k">auto</span><span class="w"> </span><span class="n">dbconn2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GenomicSQLiteOpen</span><span class="p">(</span><span class="s">&quot;compressed.db&quot;</span><span class="p">);</span><span class="w"></span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="nf">GenomicSQLiteVacuumIntoSQL</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">dest_filename</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="o">&amp;</span><span class="n">config_json</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;{}&quot;</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="n">vacuum_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GenomicSQLiteVacuumIntoSQL</span><span class="p">(</span><span class="s">&quot;compressed.db&quot;</span><span class="p">);</span><span class="w"></span>
<span class="c1">// sqlite3* dbconn opened using sqlite3_open_v2() on some existing.db</span>
<span class="c1">//   with SQLITE_OPEN_URI</span>
<span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">errmsg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">nullptr</span><span class="p">;</span><span class="w"></span>
<span class="kt">int</span><span class="w"> </span><span class="n">rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sqlite3_exec</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span><span class="w"> </span><span class="n">vacuum_sql</span><span class="p">.</span><span class="n">c_str</span><span class="p">(),</span><span class="w"> </span><span class="k">nullptr</span><span class="p">,</span><span class="w"> </span><span class="k">nullptr</span><span class="p">,</span><span class="w"> </span><span class="o">&amp;</span><span class="n">errmsg</span><span class="p">);</span><span class="w"></span>
<span class="c1">// check rc, free errmsg</span>

<span class="c1">// rc = GenomicSQLiteOpen(&quot;compressed.db&quot;, ...);</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="nf">genomicsqlite_vacuum_into_sql</span><span class="p">(</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">dest_filename</span><span class="p">,</span><span class="w"></span>
<span class="w">  </span><span class="k">const</span><span class="w"> </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">config_json</span><span class="w"></span>
<span class="p">);</span><span class="w"></span>

<span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">vacuum_sql</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">genomicsqlite_vacuum_into_sql</span><span class="p">(</span><span class="s">&quot;compressed.db&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;{}&quot;</span><span class="p">);</span><span class="w"></span>
<span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">*</span><span class="n">vacuum_sql</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="n">errmsg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">0</span><span class="p">;</span><span class="w"></span>
<span class="w">  </span><span class="cm">/* sqlite3* dbconn opened using sqlite3_open_v2() on some existing.db</span>
<span class="cm">   * with SQLITE_OPEN_URI */</span><span class="w"></span>
<span class="w">  </span><span class="kt">int</span><span class="w"> </span><span class="n">rc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sqlite3_exec</span><span class="p">(</span><span class="n">dbconn</span><span class="p">,</span><span class="w"> </span><span class="n">vacuum_sql</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="o">&amp;</span><span class="n">errmsg</span><span class="p">);</span><span class="w"></span>
<span class="w">  </span><span class="cm">/* check rc, free errmsg */</span><span class="w"></span>
<span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w"> </span><span class="cm">/* see calling convention discussed in previous examples */</span><span class="w"></span>
<span class="p">}</span><span class="w"></span>
<span class="n">sqlite3_free</span><span class="p">(</span><span class="n">vacuum_sql</span><span class="p">);</span><span class="w"></span>

<span class="cm">/* genomicsqlite_open(&quot;compressed.db&quot;, ...); */</span><span class="w"></span>
</code></pre></div>
</div>
</div>
</div>
<h3 id="genomics-extension-version">Genomics Extension version</h3>
<div class="tabbed-set tabbed-alternate" data-tabs="8:6"><input checked="checked" id="__tabbed_8_1" name="__tabbed_8" type="radio" /><input id="__tabbed_8_2" name="__tabbed_8" type="radio" /><input id="__tabbed_8_3" name="__tabbed_8" type="radio" /><input id="__tabbed_8_4" name="__tabbed_8" type="radio" /><input id="__tabbed_8_5" name="__tabbed_8" type="radio" /><input id="__tabbed_8_6" name="__tabbed_8" type="radio" /><div class="tabbed-labels"><label for="__tabbed_8_1">SQL</label><label for="__tabbed_8_2">Python</label><label for="__tabbed_8_3">Java</label><label for="__tabbed_8_4">Rust</label><label for="__tabbed_8_5">C++</label><label for="__tabbed_8_6">C</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="k">SELECT</span><span class="w"> </span><span class="n">genomicsqlite_version</span><span class="p">()</span><span class="w"></span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">genomicsqlite</span><span class="o">.</span><span class="n">__version__</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">String</span><span class="w"> </span><span class="n">genomicsqliteVersion</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GenomicSQLite</span><span class="p">.</span><span class="na">version</span><span class="p">(</span><span class="n">dbconn</span><span class="p">);</span><span class="w"></span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="kd">let</span><span class="w"> </span><span class="n">v</span>: <span class="nb">String</span> <span class="o">=</span><span class="w"> </span><span class="n">dbconn</span><span class="p">.</span><span class="n">genomicsqlite_version</span><span class="p">();</span><span class="w"></span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="n">std</span><span class="o">::</span><span class="n">string</span><span class="w"> </span><span class="nf">GenomicSQLiteVersion</span><span class="p">();</span><span class="w"></span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="kt">char</span><span class="o">*</span><span class="w"> </span><span class="nf">genomicsqlite_version</span><span class="p">();</span><span class="w"></span>
<span class="cm">/* result to be sqlite3_free() */</span><span class="w"></span>
</code></pre></div>
</div>
</div>
</div>




              
            </article>
            
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../guide_gri/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Genomic Range Indexing" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Genomic Range Indexing
            </div>
          </div>
        </a>
      
      
        
        <a href="../bindings/" class="md-footer__link md-footer__link--next" aria-label="Next: Writing Language Bindings" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Writing Language Bindings
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "..", "features": ["content.tabs.link"], "search": "../assets/javascripts/workers/search.ecf98df9.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.9c69f0bc.min.js"></script>
      
    
  </body>
</html>